phabricatoremails/source.py

```diff
@@ -0,0 +1,80 @@
+import json
+import pathlib
+from dataclasses import dataclass
+from typing import Optional
+
+import requests
+from requests import RequestException
+
+
+class PhabricatorException(Exception):
+    """An error that occurred while communicating with Phabricator"""
+
+    pass
+
+
+@dataclass
+class PhabricatorSource:
+    """Fetches feed information directly from the Phabricator API"""
+
+    server: str
+    token: str
+    override_limit: Optional[int]
+
+    def _request(self, endpoint, **query_parameters):
+        query_parameters = {k: v for k, v in query_parameters.items() if v}
+        url = "{}/api/{}".format(self.server, endpoint)
```

Comment:
Hardcoded path here -- where is `/api/` coming from?

Reply:
`/api/` is an integration detail of the Phabricator API (all endpoints are at `http://phab-host.com/api/{endpoint}`).

But, strings aren't the only integration detail of the Phabricator API. There's status code details (only `200` indicates that it's ok to proceed), HTTP details (the request could fail at a lower layer), and body details (Phabricator will define `body["error_code"]` if there's a problem.

`PhabricatorSource` is meant to encapsulate all of these integration details, so `/api/` is defined within `PhabricatorSource`.

Reply:
Ah, then I would move the base URL to an instance variable or property (e.g. `self.base_url`) and parse the endpoint url here as you are doing, to make this more clear.